transports/dns/src/lib.rs

```diff
@@ -299,28 +351,69 @@ mod tests {
             }
         }
 
-        futures::executor::block_on(async move {
-            let transport = DnsConfig::new(CustomTransport).unwrap();
-
+        async fn run<T, C, P>(transport: GenDnsConfig<T, C, P>)
+        where
+            C: DnsHandle<Error = ResolveError>,
+            P: ConnectionProvider<Conn = C>,
+            T: Transport + Clone + Send + 'static,
+            T::Error: Send,
+            T::Dial: Send,
+        {
+            // Success due to existing A record for example.com.
             let _ = transport
                 .clone()
                 .dial("/dns4/example.com/tcp/20000".parse().unwrap())
                 .unwrap()
                 .await
                 .unwrap();
 
+            // Success due to existing AAAA record for example.com.
             let _ = transport
                 .clone()
                 .dial("/dns6/example.com/tcp/20000".parse().unwrap())
                 .unwrap()
                 .await
                 .unwrap();
 
+            // Success due to pass-through, i.e. nothing to resolve.
             let _ = transport
+                .clone()
                 .dial("/ip4/1.2.3.4/tcp/20000".parse().unwrap())
                 .unwrap()
                 .await
                 .unwrap();
-        });
+
+            // Failure due to no records.
+            match transport
+                .clone()
+                .dial("/dns4/domaindoesnotexistforsureamiright.com/tcp/20000".parse().unwrap())
```

Comment:
:slightly_smiling_face: If I understand correctly an alternative would be [`invalid.`](https://en.wikipedia.org/wiki/.invalid) defined by the IETF.

Reply:
Done in https://github.com/libp2p/rust-libp2p/pull/1927/commits/37a02b9dc70e9732a4b7e6887929a0b781dd7c31.